mi/reporting/actions.py

```diff
@@ -0,0 +1,111 @@
+import csv
+import re
+from datetime import date, datetime
+from pathlib import Path
+from typing import Generator
+
+from helpers.log import log
+from mi.reporting.paths import PATH_TO_REPORT_CSV
+from mi.reporting.resources import (
+    each_sql_statement,
+    get_credentials,
+    get_endpoint,
+    get_lambda_name,
+    make_report_path,
+)
+from mi.sql_query.model import Response, Sql, SqlQueryEvent, Status
+
+SQL_SELECT_REGEX = re.compile(r"SELECT(.*)FROM", flags=re.DOTALL)
+SQL_SELECT_SEPARATOR = ","
+SQL_ALIAS_SEPARATOR_REGEX = re.compile(r"as|AS")
+
+
+@log("Created query events")
+def each_query_event(
+    session, workspace: str, env: str, partition_key
+) -> Generator[tuple[str, SqlQueryEvent], None, None]:
+    credentials = get_credentials(session=session, workspace=workspace)
+    endpoint = get_endpoint(session=session, env=env)
+    for report_name, statement in each_sql_statement():
+        yield report_name, SqlQueryEvent(
+            sql=Sql(statement=statement, params={"partition_key": partition_key}),
+            endpoint=endpoint,
+            **credentials,
+        )
+
+
+def _select_statement_from_sql_query(query: str) -> str:
+    try:
+        (select,) = SQL_SELECT_REGEX.match(query).groups()
+    except:
+        raise ValueError(f"Couldn't find valid SELECT statement in query {query}")
+    return select
+
+
+def _column_name_from_statement(column_statement: str) -> str:
+    """column_statement ~ 'foo' or 'foo as FOO' or 'foo AS FOO' --> 'foo'"""
+    parts = SQL_ALIAS_SEPARATOR_REGEX.split(column_statement)
+    last_part = parts[-1]
+    return last_part.strip()
+
+
+def _column_names_from_sql_query(query: str):
+    select = _select_statement_from_sql_query(query=query)
+    column_names = list(
+        map(_column_name_from_statement, select.split(SQL_SELECT_SEPARATOR))
+    )
+    return column_names
+
+
+@log("Querying lambda")
+def perform_query(
+    session, workspace: str, event: SqlQueryEvent
+) -> list[dict[str, any]]:
+    function_name = get_lambda_name(workspace=workspace)
+    client = session.client("lambda")
+    column_names = _column_names_from_sql_query(query=event.sql.statement)
+    raw_response = client.invoke(FunctionName=function_name, Payload=event.json())
+    response = Response.parse_raw(raw_response["Payload"].read())
+    if response.status != Status.OK:
+        raise Exception(response.outcome)
+    return [dict(zip(column_names, line)) for line in response.results]
+
+
+def _get_column_names(data: list[dict]) -> list[str]:
+    for row in data:
+        return list(row.keys())
+    return []
+
+
+@log("Wrote report to {__result__}")
+def write_csv(
+    data: list[dict],
+    env: str,
+    workspace: str,
+    report_name: str,
+    path: Path = PATH_TO_REPORT_CSV,
+    today: date = None,
```

Comment:
Im assuming these are only here for testing purposes?

Reply:
yep i'll add a comment